### ISSUE TITLE
[bug] component-hook-factories false positives due to missing equivalent static-components rule

### ISSUE THREAD

**Issue by NeoDobby** [Rel1cx/eslint-react#1632]
### Describe the bug

Below is a basic example of a react Higher Order Component, taken from https://www.w3schools.com/react//react_hoc.asp, migrated to typescript. This raises a linting error from `@eslint-react/component-hook-factories`

### Reproduction

```
import { ComponentType, ReactElement } from 'react';

// This is our HOC - it adds a border to any component
function withBorder<P>(WrappedComponent: ComponentType<P>): ComponentType<P> {
    return function NewComponent(props: any): ReactElement { // linting error raised here
        return (
            <div style={{ border: '2px solid blue', padding: '10px' }}>
                <WrappedComponent {...props} />
            </div>
        );
    };
}

interface GreetingProps {
    name: string;
}

// Simple component without border
function Greeting({ name }: GreetingProps): ReactElement {
    return <h1>Hello, {name}!</h1>;
}

// Create a new component with border
const GreetingWithBorder = withBorder<GreetingProps>(Greeting);

function App() {
    return (
        <div>
            <Greeting name="John" />
            <GreetingWithBorder name="Jane" />
        </div>
    );
}
```

### Expected behavior

This should be no problem if the HOC function is called outside of other react components, to prevent recreating the function on rerender.

My expectation is to be able to define a HOC function without any issue. A violation should be raised only if the HOC function is called inside another component, like below:

```

function App() {
    const GreetingWithBorder = withBorder<GreetingProps>(Greeting); // linting error should be raised here instead
    return (
        <div>
            <Greeting name="John" />
            <GreetingWithBorder name="Jane" />
        </div>
    );
}
```

### Platform and versions

```Text
@eslint-react/eslint-plugin@3.0.0
```

### Stack trace

```Text

```

**Comment by furcan** [comment]
Yes, 

its also false-positive complaining in tests that when the component is mocked

>  `react-x/component-hook-factories`
> 
> Do not define component 'MyComponent' inside a function. Components should be defined at the module level. Move it to the top level 

```tsx

// this
vi.mock(import('../path/to/go'), () => ({
  MyComponent: ({propX}) => <div data-x={propX} />,
}));

// or that
vi.mock(import('../path/to/go'), async (importOriginal) => {
  const whateverTheyAre = await importOriginal();
  return {
    ...whateverTheyAre,
    MyComponent: ({propX}) => <div data-x={propX} />,
  };
});
```

**Comment by Rel1cx** [comment]
We use https://react.dev/reference/eslint-plugin-react-hooks/lints/component-hook-factories as our primary reference source, rather than https://www.w3schools.com. If you think there is an issue with this rule, you should report it as an upstream issue, or first confirm that the original rule doesn't flag the provided pattern.

**Comment by Rel1cx** [comment]
> Yes,
> 
> its also false-positive complaining in tests that when the component is mocked
> 
> > `react-x/component-hook-factories`
> > Do not define component 'MyComponent' inside a function. Components should be defined at the module level. Move it to the top level
> 
> // this
> vi.mock(import('../path/to/go'), () => ({
>   MyComponent: ({propX}) => <div data-x={propX} />,
> }));
> 
> // or that
> vi.mock(import('../path/to/go'), async (importOriginal) => {
>   const whateverTheyAre = await importOriginal();
>   return {
>     ...whateverTheyAre,
>     MyComponent: ({propX}) => <div data-x={propX} />,
>   };
> });

This is the same as https://github.com/facebook/react/issues/31237#issuecomment-2706402535. Many rules should not be enabled in test files, including React's official `rules-of-hooks`. 

The fundamental issue here is **semantic pollution**: the code in test files that constructs test cases pollutes the semantic analysis of the actual source code being linted. 

You should disable these rules in your test files:

https://github.com/ginkgobioworks/react-json-schema-form-builder/blob/9d3b363a6060ecc46a108eeb9c2c49d056691490/eslint.config.mjs#L63-L68

Otherwise, every plugin author would need to implement special-case handling for every test framework (Jest, Vitest, etc.) within every rule that might be affected by test framework semantics. This approach is neither scalable nor maintainable.

**Comment by CompartMJU** [comment]
> We use https://react.dev/reference/eslint-plugin-react-hooks/lints/component-hook-factories as our primary reference source, rather than https://www.w3schools.com. If you think there is an issue with this rule, you should report it as an upstream issue, or first confirm that the original rule doesn't flag the provided pattern.

I have just recently switched from `eslint-plugin-react-hooks` to this, and I didn't remember seeing this issue in my code before. Nevertheless, I double checked with my code from above and got the following results:

* `react-hooks/component-hook-factories` isn't raised at all.
* `react-hooks/static-components` is raised in this code:

```
function App() {
    const GreetingWithBorder = withBorder<GreetingProps>(Greeting);
    return (
        <div>
            <Greeting name="John" />
            <GreetingWithBorder name="Jane" />
        </div>
    );
}
```

```
  37:14  error    Error: Cannot create components during render

Components created during render will reset their state each time they are created. Declare components outside of render.

Test.tsx:37:14
  35 |         <div>
  36 |             <Greeting name="John" />
> 37 |             <GreetingWithBorder name="Jane" />
     |              ^^^^^^^^^^^^^^^^^^ This component is created during render
  38 |         </div>
  39 |     );
  40 | }

Test.tsx:33:32
  31 |
  32 | function App() {
> 33 |     const GreetingWithBorder = withBorder<GreetingProps>(Greeting);
     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ The component is created during render here
  34 |     return (
  35 |         <div>
  36 |             <Greeting name="John" />  react-hooks/static-components
```
That rule is also not raised in my false positive example above.

So indeed, there is a difference in how the rule works between `eslint-plugin-react-hooks@7.0.1` (latest) and `@eslint-react/eslint-plugin@3.0.0` (latest).

I would update the ticket to say that a `static-components` violation should be reported instead in the appropriate cases, but I don't see that rule in your rules overview, so it's up to you how to proceed with this.

**Comment by Rel1cx** [comment]
> > We use https://react.dev/reference/eslint-plugin-react-hooks/lints/component-hook-factories as our primary reference source, rather than https://www.w3schools.com. If you think there is an issue with this rule, you should report it as an upstream issue, or first confirm that the original rule doesn't flag the provided pattern.
> 
> I have just recently switched from `eslint-plugin-react-hooks` to this, and I didn't remember seeing this issue in my code before. Nevertheless, I double checked with my code from above and got the following results:
> 
> * `react-hooks/component-hook-factories` isn't raised at all.
> * `react-hooks/static-components` is raised in this code:
> 
> ```
> function App() {
>     const GreetingWithBorder = withBorder<GreetingProps>(Greeting);
>     return (
>         <div>
>             <Greeting name="John" />
>             <GreetingWithBorder name="Jane" />
>         </div>
>     );
> }
> ```
> 
> ```
>   37:14  error    Error: Cannot create components during render
> 
> Components created during render will reset their state each time they are created. Declare components outside of render.
> 
> Test.tsx:37:14
>   35 |         <div>
>   36 |             <Greeting name="John" />
> > 37 |             <GreetingWithBorder name="Jane" />
>      |              ^^^^^^^^^^^^^^^^^^ This component is created during render
>   38 |         </div>
>   39 |     );
>   40 | }
> 
> Test.tsx:33:32
>   31 |
>   32 | function App() {
> > 33 |     const GreetingWithBorder = withBorder<GreetingProps>(Greeting);
>      |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ The component is created during render here
>   34 |     return (
>   35 |         <div>
>   36 |             <Greeting name="John" />  react-hooks/static-components
> ```
> 
> That rule is also not raised in my false positive example above.
> 
> So indeed, there is a difference in how the rule works between `eslint-plugin-react-hooks@7.0.1` (latest) and `@eslint-react/eslint-plugin@3.0.0` (latest).
> 
> I would update the ticket to say that a `static-components` violation should be reported instead in the appropriate cases, but I don't see that rule in your rules overview, so it's up to you how to proceed with this.

Thank you very much for the follow-up and the additional information. I’ll look into how to proceed and whether it’s necessary to introduce a similar `static-components` rule to reduce potential confusion caused by inconsistencies.

**Comment by furcan** [comment]
Thank you @zerone0x , @Rel1cx for the fixes 🙏
